cadence/contracts/DapperStorageRent.cdc

```diff
@@ -0,0 +1,247 @@
+  /// Attempt to refill an accounts storage capacity if it has dipped below threshold and passes other checks.
+  ///
+  /// @param address: Address to attempt a storage refill on
+  pub fun tryRefill(_ address: Address) {
+    let REFUEL_AMOUNT = 0.06;
+
+    self.cleanExpiredRefilledAccounts(10)
+
+    // Get the Flow Token reciever of the address
+    let recipient = getAccount(address)
+    let receiverRef = recipient.getCapability<&PrivateReceiverForwarder.Forwarder>(PrivateReceiverForwarder.PrivateReceiverPublicPath).borrow()
+
+    // Sliently fail if the receiverRef is nill
+    if receiverRef == nil || receiverRef!.owner == nil {
+      emit RefilledFailed(address: address, reason: "Couldn't borrow the Accounts flowTokenVault")
+      return
+    }
+
+    // Sliently fail if the account has already be refueled within the block allowance
+    if self.RefilledAccountInfos[address] != nil && getCurrentBlock().height - self.RefilledAccountInfos[address]!.atBlock < self.RefillRequiredBlocks {
+      emit RefilledFailed(address: address, reason: "RefillRequiredBlocks")
+      return
+    }
+
+    // Get the users storage capacity and usage values
+    var low: UInt64 = recipient.storageUsed
+    var high: UInt64 = recipient.storageCapacity
+
+    if high < low {
+      high <-> low
+    }
+
+    // Sliently fail if the account has been blocked from receiving refills
+    if DapperStorageRent.getBlockedAccounts().contains(address) {
+      emit RefilledFailed(address: address, reason: "Address is Blocked")
+      return
+    }
+
+    // If the user is below the threshold PrivateReceiverForwarder will send 0.06 Flow tokens for about 6MB of storage
+    if high - low < self.StorageRentRefillThreshold {
+      let vaultRef = self.account.borrow<&FlowToken.Vault>(from: /storage/flowTokenVault)
+      if vaultRef == nil {
+        emit RefilledFailed(address: address, reason: "Couldn't borrow the Accounts FlowToken.Vault")
+        return
+      }
+
+      let privateForwardingSenderRef = self.account.borrow<&PrivateReceiverForwarder.Sender>(from: PrivateReceiverForwarder.SenderStoragePath)
+      if privateForwardingSenderRef == nil {
+        emit RefilledFailed(address: address, reason: "Couldn't borrow the Accounts PrivateReceiverForwarder")
+        return
+      }
+
+      // Check to make sure the payment vault has sufficient funds
+      if let vaultBalanceRef = self.account.getCapability(/public/flowTokenBalance).borrow<&FlowToken.Vault{FungibleToken.Balance}>() {
+        if vaultBalanceRef.balance <= REFUEL_AMOUNT {
+          emit RefilledFailed(address: address, reason: "Insufficient balance to refuel")
+          return
+        }
+      } else {
+        emit RefilledFailed(address: address, reason: "Couldn't borrow flowToken balance")
+        return
+      }
+
+      // 0.06 = 6MB of storage, or ~20k NBA TS moments
+      privateForwardingSenderRef!.sendPrivateTokens(address,tokens:<-vaultRef!.withdraw(amount: REFUEL_AMOUNT))
```

Comment:
This doesn't make sure that the user would receive the 0.06 FLOW in the given account as the user can have a different receiver capability of different accounts in the private forwarder resource.

Reply:
[see note below ](https://github.com/dapperlabs/nft-catalog/pull/98#discussion_r1038377716)